src/formats/pnm.c

```diff
@@ -0,0 +1,222 @@
+// SPDX-License-Identifier: MIT
+// PNM formats decoder
+// Copyright (C) 2023 Abe Wieland <abe.wieland@gmail.com>
+
+#include "loader.h"
+
+#include <setjmp.h>
+
+// Divide a by b, rounding to the nearest integer; evaluates b twice
+#define div_near(a, b) (((a) + (b) / 2) / (b));
+// Divide a by b, rounding up; evalutes b twice
+#define div_ceil(a, b) (((a) + (b) - 1) / (b));
+
+// A file-like abstraction with error handling via setjmp/longjmp which
+// significantly helps code readability
+struct pnm_iter {
+    const uint8_t* data;
+    size_t pos;
+    size_t size;
+    jmp_buf jmp;
+};
+
+// Error conditions as passed to longjmp
+#define ERR_EOF     1
+#define ERR_RANGE   2
+#define ERR_NONDIG  3
+
+// Read a character, ignoring comments
+static uint8_t pnm_readc(struct pnm_iter* it)
+{
+    if (it->pos == it->size) longjmp(it->jmp, ERR_EOF);
+    if (it->data[it->pos] == '#') {
+        do {
+            ++it->pos;
+        } while (it->pos < it->size && it->data[it->pos] != '\n'
+                 && it->data[it->pos] != '\r');
+        if (it->pos == it->size) longjmp(it->jmp, ERR_EOF);
+    }
+    return it->data[it->pos++];
+}
+
+// Read an integer, ignoring leading whitespace
+static size_t pnm_readuint(struct pnm_iter* it)
+{
+    uint8_t c;
+    do {
+        c = pnm_readc(it);
+    } while (c == ' ' || c == '\t' || c == '\n' || c == '\r');
+    if (c < '0' || c > '9') longjmp(it->jmp, ERR_NONDIG);
+    size_t val = 0;
+    do {
```

Comment:
`strtoul`?

Reply:
Unfortunately `strtoul` is unusable for two reasons:
First, comments in PNM files may appear anywhere (even in other tokens). This fact needlessly complicates this parser, but we need to deal with it in some way to handle all possible PNM files.
Second, there isn't a guarantee the buffer is null-terminated, and `strtol` and friends don't allow a max length to be specified.

Reply:
I'm really sorry, I guess I hadn't done enough research. The standard states that comments may appear anywhere, including inside tokens; they even mention how weird this is:
> Before the whitespace character that delimits the raster, any characters from a "#" through the next carriage return or newline character, is a comment and is ignored. Note that this is rather unconventional, because a comment can actually be in the middle of what you might consider a token.

However, every parser that I've since looked at in more depth (`libnetpbm`, `gdk-pixbuf`, `libfreeimage`, and `imagemagick`) disallow comments within tokens (they simply start new tokens, as would have been the sane thing to have in the standard). This simplifies matters, though it still leaves `strtol` in an unusual spot (due to the lack of length checking); I think I'll probably still do it manually (like `libnetpbm`, `libfreeimage`, and `imagemagick`), though I could copy to an external buffer (null-terminated) and call `strtol` on that (like `gdk-pixbuf`).

Reply:
Ok, let's leave it as is.